tests/isa-test.cc

```diff
@@ -0,0 +1,718 @@
+	                    0xE7E97030), // requires e bump and T increment
+	};
+
+	void test_set_bounds()
+	{
+		int x = 0;
+		for (auto testCase : setBoundsCases)
+		{
+			testCase.run_test();
+		}
+	}
+
+	template<typename T>
+	void test_restrict_capability(Capability<T> capability, PermissionSet mask)
+	{
+		PermissionSet requestedPermissions(mask);
+		requestedPermissions &= capability.permissions();
+		PermissionSet expectedPermissions =
+		  requestedPermissions.to_representable();
+		capability.permissions() &= mask;
+		TEST(capability.permissions() == expectedPermissions,
+		     "permissions {} did not match expected {}",
+		     static_cast<PermissionSet>(capability.permissions()),
+		     expectedPermissions);
+	}
+
+	void test_and_perms()
+	{
+		int              myLocal = 0;
+		const Capability CapabilityToLocal{&myLocal};
+		const Capability CapabilityToGlobal{myGlobalIntPointer};
+		const Capability CapabilityToFunction{
+		  reinterpret_cast<const void *>(&test_and_perms)};
+		debug_log("Checking and perms results...");
+		// TODO we could do with a sealing capability too, but then this would
+		// need to become a privileged compartment.
+		// Test all possible masks on our 'root' capabilities. For now we only
+		// test permissions on data / executable capabilities. Conveniently
+		// these occupy the lower 9 bits of the permissions field.
+		for (uint32_t p = 0; p <= 0x1ff; p++)
+		{
+			auto permissionMask = PermissionSet::from_raw(p);
+			test_restrict_capability(CapabilityToLocal, permissionMask);
+			test_restrict_capability(CapabilityToGlobal, permissionMask);
+			test_restrict_capability(CapabilityToFunction, permissionMask);
+		}
+	}
+
+	[[cheri::interrupt_state(disabled)]] void *get_interrupts_disabled_sentry()
+	{
+		void *returnCap;
+		asm volatile("cjal %0, 4" : "=C"(returnCap));
+		return returnCap;
+	}
+
+	[[cheri::interrupt_state(enabled)]] void *get_interrupts_enabled_sentry()
+	{
+		void *returnCap;
+		asm volatile("cjal %0, 4" : "=C"(returnCap));
+		return returnCap;
+	}
+
+	void test_sentries()
+	{
+		// XXX the following produces a linker error:
+		// error: ld.lld: error: Compartment 'isa_test' imports library function
+		// '__library_export_isa_test__ZN12_GLOBAL__N_129get_interrupts_enabled_sentryEv'
+		// as cross-compartment call possible compiler / linker bug?
+
+		// Capability interruptsEnabledSentry = {get_interrupts_enabled_sentry};
+		// debug_log("interrupts enabled sentry {}",
+		// reinterpret_cast<void*>(interruptsEnabledSentry.get()));
+		// TEST(interruptsEnabledSentry.type() == 3,
+		//      "Expected type 3 but got {}",
+		//      interruptsEnabledSentry.type());
+
+		// Capability interruptsDisabledSentry =
+		// {&get_interrupts_disabled_sentry}; debug_log("interrupts disabled
+		// sentry {}", reinterpret_cast<void*>(interruptsDisabledSentry.get()));
+		// TEST(interruptsDisabledSentry.type() == 2,
+		//      "Expected type 2 but got {}",
+		//      interruptsDisabledSentry.type());
```

Comment:
Is there no good way to get these?

Reply:
I got the linker error in the comment. Suspect it might be a bug @davidchisnall ?

Reply:
Filed https://github.com/CHERIoT-Platform/llvm-project/issues/57